_blog/2024-01-24-using-cronjobs-rust.mdx

```diff
@@ -0,0 +1,294 @@
+
+```rust
+#[derive(Clone)]
+struct CronService { 
+    message: String 
+}
+
+impl CronService {
+    fn execute(&self, item: Reminder) {
+        println!("{} from CronService::execute()!", &self.message);
+    }
+}
+```
+
+You can then feed this struct back into the `say_hello_world` function, which uses the `execute()` function:
+
+```rust
+async fn say_hello_world(job: Reminder, ctx: JobContext) {
+    println!("Hello world from send_reminder()!");
+    // this lets you use variables stored in the CronService struct
+    let svc = ctx.data_opt::<CronService>().unwrap();
+    // this executes CronService::execute()
+    svc.execute(job);
+}
+```
+
+### Hooking it all up
+
+Now we’re going to hook the structs we made previously into our main app!
+
+To start with, we need to set up the `PostgresStorage` type so that we can use Postgres for durable job queues. Without durable job queues, our jobs would disappear if our web service has any outages! We can convert this directly from the `PgPool` type stored in our main struct and run the migrations for it like so:
+
+```rust
+#[shuttle_runtime::async_trait]
+impl shuttle_runtime::Service for MyService {
+    async fn bind(
+        self, 
+        _addr: std::net::SocketAddr
+    ) -> Result<(), shuttle_runtime::Error> {
+        // set up Postgres-backed storage
+        let storage = PostgresStorage::new(self.db);
+        // set up storage
+        storage.setup().await.expect("Unable to run migrations :(");
+    Ok(())
+   }
+}
+```
+
+Now we can create a `tower` service that holds our job that we want to run. We can use the `DefaultRetryPolicy` provided to us by `apalis` to be able to add automatic re-tries, as well as layering the shared data as an `Extension`:
+
+```rust
+use tower::ServiceBuilder;
+use apalis::layers::{DefaultRetryLayer, Extension, RetryLayer};
+use apalis::prelude::job_fn;
+
+#[shuttle_runtime::async_trait]
+impl shuttle_runtime::Service for MyService {
+    async fn bind(
+        self, 
+        _addr: std::net::SocketAddr
+    ) -> Result<(), shuttle_runtime::Error> {
+    // .. your previous code
+
+    let cron_service_ext = CronService {
+    message: "Hello world".to_string(),
+};
+
+// create a servicebuilder for the cronjob
+let service = ServiceBuilder::new()
+    .layer(RetryLayer::new(DefaultRetryPolicy))
+    .layer(Extension(cron_service_ext))
+    .service(job_fn(say_hello_world));
+    }
+}
+```
+
+Now that we’ve built the service, we need to build the worker and then finally create an `apalis::Monitor` process that will carry the work out:
+
+```rust
+use apalis::prelude::timer::TokioTimer;
+use apalis::cron::{Schedule, CronStream};
+use apalis::prelude::{WorkerBuilder, Monitor};
+
+#[shuttle_runtime::async_trait]
+impl shuttle_runtime::Service for MyService {
```

Comment:
How can you implement twice the same trait for the same struct ?

Reply:
Ah it's actually in the same part. I would display it differently then, this is confusing otherwise. The comment `// .. your previous code` can easily be missed.